### PR TITLE
Inject FilterPluginManager and FormatterPluginManager into AbstractWriter

### PR DESCRIPTION
**Pull request by waltertamboer**
This PR makes it possible to inject the FilterPluginManager and the FormatterPluginManager into writers.

This PR fixese issue #36

### CODE REVIEW

**Comment by waltertamboer** [comment]
I think all feedback has been processed. Could you do a review again?

File: src/Writer/Factory/WriterFactory.php [hunk 1]
```diff
@@ -0,0 +1,126 @@
+<?php
+/**
+ * Zend Framework (http://framework.zend.com/)
+ *
+ * @link      http://github.com/zendframework/zf2 for the canonical source repository
+ * @copyright Copyright (c) 2005-2015 Zend Technologies USA Inc. (http://www.zend.com)
```
Comment (samsonasik) [inline src/Writer/Factory/WriterFactory.php:6]: 2016

Comment (waltertamboer) [inline src/Writer/Factory/WriterFactory.php:6]: Thanks @samsonasik You are right. If you do not mind I'd like to suggest that this will be fixed in a separate PR. The docblocks are wrong in every file and I don't want to make this PR too big.

There's also an issue about this: #47

Comment (samsonasik) [inline src/Writer/Factory/WriterFactory.php:6]: ok

File: src/Writer/AbstractWriter.php [hunk 2]
```diff
@@ -77,6 +79,14 @@ public function __construct($options = null)
         }
 
         if (is_array($options)) {
+            if (isset($options['filter_manager']) && $options['filter_manager'] instanceof FilterPluginManager) {
+                $this->setFilterPluginManager($options['filter_manager']);
+            }
+
+            if (isset($options['formatter_manager']) && $options['formatter_manager'] instanceof FormatterPluginManager) {
```
Comment (Xerkus) [inline src/Writer/AbstractWriter.php]: mmm, how about skipping instanceof check and letting it to hard fail on setter typehint if unexpected value is provided? That will close off potential for pretty big wtf moment.

Comment (waltertamboer) [inline src/Writer/AbstractWriter.php]: Sure, makes sense, will update it.

**Comment by waltertamboer** [comment]
@Xerkus Fixed.

**Comment by Xerkus** [comment]
:+1: looks good

File: doc/book/service-manager.md [hunk 3]
```diff
@@ -0,0 +1,99 @@
+# Service Manager
+
+The `zend-log` package provides several components which can be used to
```
Comment (Xerkus) [inline doc/book/service-manager.md]: > used **to** in

looks wrong

File: doc/book/service-manager.md [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+in combination with the service managaer. These components make it possible
```
Comment (Xerkus) [inline doc/book/service-manager.md]: > manag**a**er

**Comment by waltertamboer** [comment]
Thanks @Xerkus I've fixed the typo's.

**Comment by Xerkus** [comment]
Hm. I think documentation should be focused on standalone usage, while covering zend-mvc as special case
I'll try to write some tomorrow

**Comment by waltertamboer** [comment]
@Xerkus I've had a short chat with @weierophinney about this on IRC. Maybe you guys need align thoughts on this so we're all on the same page.

**Comment by Xerkus** [comment]
Hm, i see plugin managers initialization was moved from ServiceListenerFactory in zend-mvc to Module class in zend-log

@waltertamboer what i meant is to expand docs to cover what is needed to use zend-log standalone and then "if you use zend-mvc(zend-modulemanager?), this and that is already provided for your convenience"
Say, if i try to follow proposed docs with zend-expressive, it just won't work since there is no module manager to provide ServiceListener or to invoke https://github.com/zendframework/zend-log/blob/master/src/Module.php#L32
Actually, i think it is broad enough topic for separate issue. No more objections here.

@weierophinney LGTM

**Comment by waltertamboer** [comment]
What is the hold-up? Can I do anything?

**Comment by weierophinney** [comment]
@waltertamboer I'm the hold-up; busy trying to finalize tasks for the ZF3 release, and this dropped off my radar. I'll review soon; thanks for working on this!

File: doc/book/service-manager.md [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+```php
+// module.config.php
+return [
+    'service_manager' => [
+        'abstract_factories' => [
+            'Zend\Log\LoggerAbstractServiceFactory',
+        ],
+    ],
+];
+```
```
Comment (samsonasik) [inline doc/book/service-manager.md:23]: As `Zend\Log` is now a module that provide ConfigProvider, and can be registered if using `ZendSkeletonApplication`, I think we can add note:

If you are using ZendSkeletonApplication, the `LoggerAbstractServiceFactory` will automatically registered by registering `Zend\Log` into `config/application.config.php`:

``` php
return array(
    'modules' => array(
        'Zend\Log',
     ),
    //  .... 
);
```

Comment (weierophinney) [inline doc/book/service-manager.md:23]: I can make that change on merge.

@waltertamboer What @samsonasik is saying is that the registration of the abstract factory is only necessary in ZF2 applications; once you bump to v3 and/or start using zend-component-installer, the registration happens automatically for you. What I'll do on merge is add version-specific verbiage.

File: src/Writer/Factory/WriterFactory.php [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+     * @param null|array
+     */
+    private $creationOptions;
+
+    /**
+     * @param null|array|Traversable $creationOptions
+     * @throws InvalidServiceException if $creationOptions cannot be coerced to
+     *     an array.
+     */
+    public function __construct($creationOptions = null)
```
Comment (weierophinney) [inline src/Writer/Factory/WriterFactory.php]: Typehint `$creationOptions` as an array.

File: src/Writer/Factory/WriterFactory.php [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+    {
+        if (null === $creationOptions) {
+            return;
+        }
+
+        $this->setCreationOptions($creationOptions);
```
Comment (weierophinney) [inline src/Writer/Factory/WriterFactory.php]: Invert these to:

``` php
if (is_array($creationOptions)) {
    $this->setCreationOptions($creationOptions);
}
```

(I can do this on merge, if I get done with the review today)

File: src/Writer/Factory/WriterFactory.php [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+     *
+     * @param ServiceLocatorInterface $serviceLocator
+     * @param null|string $canonicalName
+     * @param null|string $requestedName
+     * @return object
+     * @throws InvalidServiceException
+     */
+    public function createService(ServiceLocatorInterface $serviceLocator, $canonicalName = null, $requestedName = null)
+    {
+        if (class_exists($canonicalName)) {
```
Comment (weierophinney) [inline src/Writer/Factory/WriterFactory.php]: Should test that `$canonicalName` is a string first, as you do with `$requestedName`.

Comment (waltertamboer) [inline src/Writer/Factory/WriterFactory.php]: I'll update this. I've copied this class from some other zend-\* package. This might be wrong in multiple libraries.

File: src/Writer/Factory/WriterFactory.php (hunk 7 above)
Comment (waltertamboer) [inline src/Writer/Factory/WriterFactory.php]: Done

File: src/Writer/Factory/WriterFactory.php [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+     * @param ContainerInterface $container
+     * @param string $requestedName
+     * @param null|array $options
+     * @return object
+     */
+    public function __invoke(ContainerInterface $container, $requestedName, array $options = null)
+    {
+        if (isset($options['filter_manager']) && is_string($options['filter_manager'])) {
+            $options['filter_manager'] = $container->get($options['filter_manager']);
+        } elseif (!isset($options['filter_manager']) && $container->has('LogFilterManager')) {
```
Comment (weierophinney) [inline src/Writer/Factory/WriterFactory.php]: We likely want to grab the canonical `LogFilterManager` service (and the `LogFormatterManager` service, from below) if `$options['filter_manager']` is set but not a string and/or the named service does not exist. I'd move that logic into a separate method, something like this:

``` php
private function getFilterManager(ContainerInterface $container, array $options)
{
    if (! isset($options['filter_manager']) && ! $container->has('LogFilterManager')) {
        return null;
    }

    if (! isset($options['filter_manager']) && $container->has('LogFilterManager')) {
    return $container->get('LogFilterManager');
    }

    if (isset($options['filter_manager'])
        && is_string($options['filter_manager'])
        && $container->has($options['filter_manager'])
    ) {
        return $container->get($options['filter_manager']);
    }

    throw new ServiceNotCreatedException('Invalid zend-log filter_manager option provided');
}
```

which you would then call from `__invoke()` using:

``` php
$options = (array) $options;
if (null !== ($filterManager = $this->getFilterManager($container, $options))) {
    $options['filter_manager'] = $filterManager;
}
```

You would then do the same with the formatter manager. This approach ensures that we have valid arguments, and allows us to raise a very specific exception about why we cannot create the _requested_ service (invalid filter/formatter configuration).

File: src/Writer/Factory/WriterFactory.php [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+     * Create an instance of the requested class name.
+     *
+     * @param ContainerInterface $container
+     * @param string $requestedName
+     * @param null|array $options
+     * @return object
+     */
+    public function __invoke(ContainerInterface $container, $requestedName, array $options = null)
+    {
+        if (isset($options['filter_manager']) && is_string($options['filter_manager'])) {
```
Comment (weierophinney) [inline src/Writer/Factory/WriterFactory.php]: Need to cast `$options` to an array before testing keys.

File: src/Writer/Factory/WriterFactory.php [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+            $options['filter_manager'] = $container->get('LogFilterManager');
+        }
+
+        if (isset($options['formatter_manager']) && is_string($options['formatter_manager'])) {
+            $options['formatter_manager'] = $container->get($options['formatter_manager']);
+        } elseif (!isset($options['formatter_manager']) && $container->has('LogFormatterManager')) {
+            $options['formatter_manager'] = $container->get('LogFormatterManager');
+        }
+
+        return (null === $options) ? new $requestedName : new $requestedName($options);
```
Comment (weierophinney) [inline src/Writer/Factory/WriterFactory.php]: I'd argue we do this at the start of the method:

``` php
if (null === $options) {
    return new $requestedName;
}
```

That serves the purpose of ensuring that when we start testing for array keys of `$options`, we have an array, and then removes this ternary (it becomes `return new $requestedName($options);`).

File: src/Writer/Factory/WriterFactory.php (hunk 10 above)
Comment (waltertamboer) [inline src/Writer/Factory/WriterFactory.php]: Good catch, I'll add a line `$options = (array)$options;`

File: test/Writer/AbstractTest.php [hunk 12]
```diff
@@ -125,6 +128,78 @@ public function testConstructorWithPriorityFilter()
...
+        // Arrange
+        // ...
+
+        // Act
+        $writer = new ConcreteWriter([
+            'formatter_manager' => 123,
+        ]);
+
+        // Assert
+        $this->assertInstanceOf(FormatterPluginManager::class, $writer->getFormatterPluginManager());
```
Comment (weierophinney) [inline test/Writer/AbstractTest.php]: Is this assertion even necessary? The exception will be raised during instantiation...

File: test/Writer/AbstractTest.php [hunk 13]
```diff
@@ -125,6 +128,78 @@ public function testConstructorWithPriorityFilter()
...
+        // Arrange
+        // ...
+
+        // Act
+        $writer = new ConcreteWriter([
+            'filter_manager' => 123,
+        ]);
+
+        // Assert
+        $this->assertInstanceOf(FilterPluginManager::class, $writer->getFilterPluginManager());
```
Comment (weierophinney) [inline test/Writer/AbstractTest.php]: Same comment as with the formatter plugin manager.

File: src/Writer/Factory/WriterFactory.php [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+            $options['filter_manager'] = $container->get('LogFilterManager');
+        }
+
+        if (isset($options['formatter_manager']) && is_string($options['formatter_manager'])) {
+            $options['formatter_manager'] = $container->get($options['formatter_manager']);
+        } elseif (!isset($options['formatter_manager']) && $container->has('LogFormatterManager')) {
+            $options['formatter_manager'] = $container->get('LogFormatterManager');
+        }
+
+        return new $requestedName($options);
```
Comment (waltertamboer) [inline src/Writer/Factory/WriterFactory.php:63]: @weierophinney I've changed this line as well since I'm now casting $options to an array. The old line was `return (null === $options) ? new $requestedName : new $requestedName($options);`

Do you agree?

File: test/Writer/Factory/WriterFactoryTest.php [hunk 15]
```diff
@@ -0,0 +1,290 @@
...
+
+class WriterFactoryTest extends TestCase
+{
+    /**
+     * @covers \Zend\Log\Writer\Factory\WriterFactory::setCreationOptions
+     */
+    public function testSetCreationOptions()
+    {
+        // Arrange
+        $container = $this->getMockForAbstractClass(ServiceLocatorInterface::class);
```
Comment (weierophinney) [inline test/Writer/Factory/WriterFactoryTest.php]: You're mixing PHPUnit mocks with Prophecy mocks; use one or the other. This one can become:

``` php
$container = $this->prophesize(ServiceLocatorInterface::class);
$container->willImplement(ContainerInterface::class);
$container = $container->reveal();
```

File: test/Writer/Factory/WriterFactoryTest.php [hunk 16]
```diff
@@ -0,0 +1,290 @@
...
+        $pluginManager->getServiceLocator()->willReturn($container);
+        $pluginManager = $pluginManager->reveal();
+
+        $factory = new WriterFactory();
+
+        // Act
+        $object = $factory->createService($pluginManager, 'invalid', 'invalid');
+
+        // Assert
+        $this->assertInstanceOf(InvokableObject::class, $object);
```
Comment (weierophinney) [inline test/Writer/Factory/WriterFactoryTest.php]: The exceptions will be raised before this.

This is one place where I wouldn't use the annotations, but instead call `setExpectedException()` immediately prior to the call that will raise the exception; this makes it clear when setup code raises an exception vs when execution code raises it. As such:

``` php
$this->setExpectedException(
    InvalidServiceException::class,
    'WriterFactory requires that the requested name is provided'
);
$factory->createService($pluginManager, 'invalid', 'invalid');
```

File: test/Writer/Factory/WriterFactoryTest.php [hunk 17]
```diff
@@ -0,0 +1,290 @@
...
+     */
+    public function testInvokeWithInvalidFilterManagerAsString()
+    {
+        // Arrange
+        $container = $this->getMockForAbstractClass(ContainerInterface::class);
+        $factory = new WriterFactory();
+
+        // Act
+        $object = $factory($container, InvokableObject::class, [
+            'filter_manager' => 'my_manager',
```
Comment (weierophinney) [inline test/Writer/Factory/WriterFactoryTest.php:169]: If we make the changes as I suggested in the WriterFactory class above, this use case will now raise an exception (as an invalid manager is specified); this test will then need to change to assert against the expected exception.

File: test/Writer/Factory/WriterFactoryTest.php [hunk 18]
```diff
@@ -0,0 +1,290 @@
...
+     */
+    public function testInvokeWithInvalidFormatterManagerAsString()
+    {
+        // Arrange
+        $container = $this->getMockForAbstractClass(ContainerInterface::class);
+        $factory = new WriterFactory();
+
+        // Act
+        $object = $factory($container, InvokableObject::class, [
+            'formatter_manager' => 'my_manager',
```
Comment (weierophinney) [inline test/Writer/Factory/WriterFactoryTest.php:235]: Same comment as for the filter manager; invalid string should raise an exception, as it's an invalid service.

File: src/Writer/Factory/WriterFactory.php (hunk 14 above)
Comment (weierophinney) [inline src/Writer/Factory/WriterFactory.php:63]: That works too. :)

Comment (weierophinney) [inline src/Writer/Factory/WriterFactory.php:63]: I'd still do the filter_manager/formatter_manager validation in separate methods, raising an exception for invalid values, though.
